StreamChatClient_v3/ChatClient.swift

```diff
@@ -0,0 +1,85 @@
+//
+// ChatClient.swift
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+public final class ChatClient {
+  // MARK: - Public
+
+  public let currentUser: User
+
+  public let config: Config
+
+  public convenience init(currentUser: User, config: ChatClient.Config = .init()) {
+    // All production workers
+    let workers: [WorkerBuilder] = [
+      MessageSender.init
+    ]
+
+    self.init(
+      currentUser: currentUser,
+      config: config,
+      workers: workers,
+      environment: .init()
+    )
+  }
+
+  // MARK: - Internal
+
+  struct Environment {
+    var apiClientBuilder: (_ currentUser: User) -> APIClient = APIClient.init
+    var webSocketClientBuilder: (_ currentUser: User) -> WebSocketClient = WebSocketClient.init
+  }
+
+  var workers: [Worker]! // TODO: make `let` or `private(set) lazy var`
+
+  private let environment: Environment
+
+  private lazy var apiClient: APIClient = self.environment.apiClientBuilder(self.currentUser)
+
+  private lazy var webSocketClient: WebSocketClient = self.environment.webSocketClientBuilder(self.currentUser)
+
+  private lazy var persistentContainer: NSPersistentContainer = {
```

Comment:
I think this should be in `Storage`. We want to make Client thin, but you add more responsibility. The `Client` shouldn't know about CoreData.

Reply:
Comming in CIS-165